markdown_it/tree.py

```diff
@@ -35,17 +49,17 @@ def __init__(self) -> None:
         self.token: Optional[Token] = None
 
         # Only containers have nester tokens
-        self.nester_tokens: Optional[SyntaxTreeNode._NesterTokens] = None
+        self.nester_tokens: Optional[_NesterTokens] = None
 
         # Root node does not have self.parent
-        self.parent: Optional["SyntaxTreeNode"] = None
+        self.parent = None
 
         # Empty list unless a non-empty container, or unnested token that has
         # children (i.e. inline or img)
-        self.children: List["SyntaxTreeNode"] = []
+        self.children = []
 
     @classmethod
-    def from_tokens(cls, tokens: Sequence[Token]) -> "SyntaxTreeNode":
+    def from_tokens(cls: Type[_T], tokens: Sequence[Token]) -> _T:
```

Comment:
do `cls` and `self` really need to be annotated?

Reply:
Yep, the type checker binds a meaning to `_T` based on context. Here it gets the context from `cls` (it knows its type) so it deduces that the return type is the same as that.

E.g. if you call `node = MyTreeNode.from_tokens()`, the type checker then knows that `node` is `MyTreeNode` and if the type is something different (another subclass) the type checker binds the `TypeVar` to that.

Reply:
ok cheers

Reply:
To clarify, if you dont annnotate `cls`, the type checker doesn't have any context useful to binding the `TypeVar` so the return type will be `Any`.